src/certstatus/certstatus.c

```diff
@@ -416,3 +416,159 @@ int check_revocation_any_method(X509_STORE_CTX* ctx)
+    int res = (*check_revocation)(tmp_ctx); /* checks only depth 0 */
+
+    if(res is_eq 0)
+    {
+        ret = X509_STORE_CTX_get_error(tmp_ctx);
+    }
+    else
+    {
+        ret = X509_V_OK;
+    }
+
+ err:
+    X509_STORE_CTX_free(tmp_ctx);
+    return ret;
+}
+
+static bool crl_time_valid(const X509_CRL* crl, const X509_VERIFY_PARAM* vpm)
+{
+    time_t check_time, *ptime = 0;
+    const ASN1_TIME* crl_end_time;
+    const ASN1_TIME* crl_last_update;
+    unsigned long flags = X509_VERIFY_PARAM_get_flags((X509_VERIFY_PARAM*)vpm);
+
+    if((flags bitand X509_V_FLAG_USE_CHECK_TIME) not_eq 0)
+    {
+        check_time = X509_VERIFY_PARAM_get_time(vpm);
+        ptime = &check_time;
+    }
+    crl_end_time = X509_CRL_get0_nextUpdate(crl);
+    crl_last_update = X509_CRL_get0_lastUpdate(crl);
+
+    if ((crl_end_time not_eq 0 and X509_cmp_time(crl_end_time, ptime) not_eq 1) or
+        (crl_last_update not_eq 0 and X509_cmp_time(crl_last_update, ptime) not_eq -1))
+    {
+        return false;
+    }
+
+    return true;
+}
+
+int check_revocation_local_only_method(X509_STORE_CTX* ctx)
+{
+    if(0 is_eq ctx)
+    {
+        LOG(FL_ERR, "null parameter ctx");
+        return 0;
+    }
+
+    STACK_OF(X509) *chain = X509_STORE_CTX_get0_chain(ctx);
+    int i = 0;
+    const int last = sk_X509_num(chain) - 1;
+
+    for(i = 0; i <= last; i++)
+    {
+        X509* cert = sk_X509_value(chain, i);
+        if(i is_eq last and X509_check_issued(cert, cert) is_eq X509_V_OK)
+        {
+            LOG_cert(FL_DEBUG, "skipping revocation check for self-issued last", cert);
+            break;
+        }
+
+        STACK_OF(X509_CRL)* crls = X509_STORE_CTX_get1_crls(ctx, X509_get_issuer_name(cert));
+        if(not crls or 0 is_eq sk_X509_CRL_num(crls) or 0 is_eq sk_X509_CRL_value(crls, 0))
+        {
+            char* issuer = X509_NAME_oneline(X509_get_issuer_name(cert), 0, 0);
+            if (0 not_eq issuer)
+            {
+                LOG(FL_ERR, "Local CRL for %s certificate %s not found",
+                    (i + 1 is_eq last) ? "root" : "issuing", issuer);
+                OPENSSL_free(issuer);
+            }
+            return 0;
+        }
+
+        if(true not_eq crl_time_valid(sk_X509_CRL_value(crls, 0), X509_STORE_CTX_get0_param(ctx)))
+        {
+            char* issuer = X509_NAME_oneline(X509_CRL_get_issuer(sk_X509_CRL_value(crls, 0)), 0, 0);
+            if(issuer not_eq 0)
+            {
+                LOG(FL_WARN, "CRL issued by %s is not valid in time", issuer);
+                OPENSSL_free(issuer);
+            }
+        }
+
+        sk_X509_CRL_pop_free(crls, X509_CRL_free);
+
+        X509_STORE_CTX_set_error_depth(ctx, i);
+        int res = check_local_cert_crls(ctx);
+
+        if(X509_V_OK not_eq res)
+        {
+            verify_cb_cert(ctx, cert, res);
+            return 0;
+        }
+        chain = X509_STORE_CTX_get0_chain(ctx); /* for some reason need again */
+    }
+    return 1;
+}
```

Comment:
Sorry for all the effort you spent here, but I see no point in defining `check_revocation_local_only_method()` and `STORE_load_crl_dir_local_only()` because
`check_revocation_any_method()` should already do all you need if no extra cert status sources (via CDPs or OCSP) are defined or enabled, doesn't it?

And even if not, it would have been much better not to try to copy and strip down `check_revocation_any_method()`, which just causes much redundancy and very likely introduces errors and omissions, but to adapt `check_revocation_any_method()` - if this is needed after all.